src/Components/EnvironmentPicker/EnvironmentPicker.test.tsx

```diff
@@ -0,0 +1,216 @@
+            expect(result.storageAccountItems.storageAccounts.length).toEqual(
+                1
+            );
+            expect(
+                result.storageAccountItems.storageAccountToEdit
+            ).toMatchObject(MOCK_STORAGE_ACCOUNT);
+        });
+
+        test('[SET_CONTAINER_ITEMS] - setting the storage container related items in the state', () => {
+            // ARRANGE
+            const initialState = MOCK_ENVIRONMENT_STATE;
+            initialState.containerItems = {
+                containers: [],
+                containerToEdit: null
+            }; // no items
+
+            const action: EnvironmentPickerAction = {
+                type: EnvironmentPickerActionType.SET_CONTAINER_ITEMS,
+                payload: {
+                    containerItems: {
+                        containers: [MOCK_STORAGE_CONTAINER],
+                        containerToEdit: MOCK_STORAGE_CONTAINER
+                    }
+                }
+            };
+
+            // ACT
+            const result = EnvironmentPickerReducer(initialState, action);
+
+            // ASSERT
+            expect(result.containerItems.containers.length).toEqual(1);
+            expect(result.containerItems.containerToEdit).toMatchObject(
+                MOCK_STORAGE_CONTAINER
+            );
+        });
+
+        test('[SET_FIRST_TIME_VISIBLE] - setting the first time visibility of dialog in the state', () => {
+            // ARRANGE
+            const initialState = MOCK_ENVIRONMENT_STATE;
+            initialState.firstTimeVisible = false;
+
+            const action: EnvironmentPickerAction = {
+                type: EnvironmentPickerActionType.SET_FIRST_TIME_VISIBLE,
+                payload: true
+            };
+
+            // ACT
+            const result = EnvironmentPickerReducer(initialState, action);
+
+            // ASSERT
+            expect(result.firstTimeVisible).toBeTruthy();
+        });
+
+        test('[RESET_ITEMS_ON_DISMISS] - resetting all the items of the state back to selected values after dismissing the modal', () => {
+            // ARRANGE
+            const initialState = MOCK_ENVIRONMENT_STATE;
+            initialState.environmentItems = {
+                environments: [MOCK_ENVIRONMENT, MOCK_ENVIRONMENT2],
+                environmentToEdit: MOCK_ENVIRONMENT2
+            };
+            initialState.storageAccountItems = {
+                storageAccounts: [MOCK_STORAGE_ACCOUNT, MOCK_STORAGE_ACCOUNT2],
+                storageAccountToEdit: MOCK_STORAGE_ACCOUNT2
+            };
+            initialState.containerItems = {
+                containers: [MOCK_STORAGE_CONTAINER, MOCK_STORAGE_CONTAINER2],
+                containerToEdit: MOCK_STORAGE_CONTAINER2
+            };
+
+            const action: EnvironmentPickerAction = {
+                type: EnvironmentPickerActionType.RESET_ITEMS_ON_DISMISS,
+                payload: {
+                    selectedEnvironmentUrl: MOCK_ENVIRONMENT_URL,
```

Comment:
why does a reset operation require the consumer to pass in all the original data? Shouldn't it know the previous state internally and manage that for the consumer? Seems like it's just a massive "Set" action rather than a "Reset"

Reply:
what do you mean by "consumer"? The component knows the original selected data through its prop and I just send the value of it to the reducer to update the internal state variables. Yes it is basically "re-setting" the options back to its originals by selected values passed by props. In internal state we only keep the draft values which is subject to change (e.g. including deleting the currently selected freeform option), but when we dismiss the dialog we want to reset the selected options back to the selected values passed thought props. Since there are many things to check and reset it should be a separate reducer action to reset the variables in state.